process/notifier/sovereignNotifier.go

```diff
@@ -1,33 +1,199 @@
+func getAddressesMap(addresses [][]byte) (map[string]struct{}, error) {
+	numAddresses := len(addresses)
+	if numAddresses == 0 {
+		return nil, errNoSubscribedAddresses
+	}
+
+	addressesMap := make(map[string]struct{})
+	for _, addr := range addresses {
+		addressesMap[string(addr)] = struct{}{}
+	}
+
+	if len(addressesMap) != numAddresses {
+		return nil, errDuplicateSubscribedAddresses
+	}
+
+	return addressesMap, nil
+}
+
+// Notify will notify the sovereign nodes about the finalized block and incoming mb txs
+// For each subscribed address, it searches if the receiver is found in tx in pool.
+// If found, IncomingMiniBlocks will contain the ordered tx hashes by execution.
+func (notifier *sovereignNotifier) Notify(outportBlock *outport.OutportBlock) error {
+	err := checkNilOutportBlockFields(outportBlock)
+	if err != nil {
+		return err
+	}
+
+	mbs, err := notifier.getAllIncomingMbs(outportBlock.TransactionPool)
+	if err != nil {
+		return err
+	}
+
+	headerV2, err := notifier.getHeaderV2(core.HeaderType(outportBlock.BlockData.HeaderType), outportBlock.BlockData.HeaderBytes)
+	if err != nil {
+		return err
+	}
+
+	extendedHeader := &block.ShardHeaderExtended{
+		Header:             headerV2,
+		IncomingMiniBlocks: mbs,
+	}
+
+	notifier.notifyHandlers(extendedHeader)
+	return nil
+}
+
+func checkNilOutportBlockFields(outportBlock *outport.OutportBlock) error {
+	if outportBlock == nil {
+		return errNilOutportblock
+	}
+	if outportBlock.TransactionPool == nil {
+		return errNilTransactionPool
+	}
+	if outportBlock.BlockData == nil {
+		return errNilBlockData
+	}
+
+	return nil
+}
+
+func (notifier *sovereignNotifier) getAllIncomingMbs(txPool *outport.TransactionPool) ([]*block.MiniBlock, error) {
+	mbs := make([]*block.MiniBlock, 0)
+
+	txsMb, err := notifier.getIncomingMbFromTxs(txPool.Transactions)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: when specs are defined, we should also handle scrs mbs
+	mbs = append(mbs, txsMb)
+	return mbs, nil
+}
+
+func (notifier *sovereignNotifier) getIncomingMbFromTxs(txs map[string]*outport.TxInfo) (*block.MiniBlock, error) {
+	txHashes := make([][]byte, 0)
+	execOrderTxHashMap := make(map[string]uint32)
+
+	for txHash, tx := range txs {
+		receiver := tx.GetTransaction().GetRcvAddr()
+		_, found := notifier.subscribedAddresses[string(receiver)]
+		if !found {
+			continue
+		}
+
+		hashBytes, err := hex.DecodeString(txHash)
+		if err != nil {
+			return nil, fmt.Errorf("%w, hash: %s", err, txHash)
+		}
+
+		log.Info("found incoming tx", "tx hash", txHash, "address", hex.EncodeToString(receiver))
+
+		txHashes = append(txHashes, hashBytes)
+		execOrderTxHashMap[string(hashBytes)] = tx.GetExecutionOrder()
+	}
+
+	sort.SliceStable(txHashes, func(i, j int) bool {
+		return execOrderTxHashMap[string(txHashes[i])] < execOrderTxHashMap[string(txHashes[j])]
+	})
+
+	return &block.MiniBlock{
+		TxHashes:        txHashes,
+		ReceiverShardID: 0, // todo: decide what we should fill here
```

Comment:
The ReceiverShardID could be core.SovereignShardId and the SenderShardID could be the real sender shard . But we will decide later.

Reply:
Filled `ReceiverShardID` with `core.SovereignShardId`
Added `ShardCoordinator` inside `Notifier` to be used in a future PR to compute this field and create multiple mbs,e.g.:
- `Mb_from_sh0_to_sovereign`
- `Mb_from_sh1_to_sovereign`
....